test/schema-special-usage.test.js

```diff
@@ -47,6 +50,122 @@ test('Ajv8 usage instead of the bundle one', t => {
+                version: { const: '2' },
+                foo: { type: 'number' }
+              },
+              required: ['version']
+            }
+          ]
+        }
+      },
+      handler (req, reply) { reply.send({ ok: 1 }) }
+    })
+
+    fastify
+      .ready(err => {
+        t.ok(err)
+        t.match(
+          err ? err.message : undefined,
+          `Failed building the validation schema for POST: /,
+           due to error schema is invalid: data/discriminator must NOT have additional properties,
+           data/oneOf must NOT have additional properties,
+           data must have property 'ref',
+           data/discriminator must NOT have additional properties,
+           data/oneOf must NOT have additional properties,
+           data must have property 'type',
+           data/discriminator must NOT have additional properties,
+           data/oneOf must NOT have additional properties,
+           data must have property 'enum',
+           data/discriminator must NOT have additional properties,
+           data/oneOf must NOT have additional properties,
+           data must have property 'elements',
+           data/discriminator must NOT have additional properties,
+           data/oneOf must NOT have additional properties,
+           data must have property 'properties',
+           data/discriminator must NOT have additional properties,
+           data/oneOf must NOT have additional properties,
+           data must have property 'optionalProperties',
+           data/discriminator must NOT have additional properties,
+           data/oneOf must NOT have additional properties,
+           data/discriminator must be string,
+           data must have property 'mapping',
+           data/oneOf must NOT have additional properties,
+           data must have property 'values',
+           data/discriminator must NOT have additional properties,
+           data/oneOf must NOT have additional properties,
+           data must match a schema in union
+          `.replace(/\n|( {2})+/g, ''),
+          'the JTD schema won\'t compile when using strict mode JSON Schema'
+        )
+      })
+  })
+
+  t.test('use JTD explicit substitution for basic JTD support', t => {
+    t.plan(1)
+    const ajv = new AjvJTD({
+      allErrors: true
+    })
+    const fastify = Fastify({ jsonShorthand: false })
+    fastify.setValidatorCompiler(({ schema }) => {
+      return ajv.compile(schema)
+    })
+
+    fastify.post('/', {
+      schema: {
+        body: {
+          discriminator: 'version',
+          mapping: {
+            1: {
+              properties: {
+                foo: { type: 'uint8' }
+              }
+            },
+            2: {
+              properties: {
+                foo: { type: 'string' }
+              }
+            }
+          }
+        }
+      },
+      handler (req, reply) { reply.send({ ok: 1 }) }
+    })
+
+    fastify
+      .ready(err => {
+        t.equal(err, undefined)
```

Comment:
I'm not a fan of this test as said, but in case i would improve it with

```

fastify.inject({
  method: 'POST',
  url:'/',
  payload:{
    foo:'asd',
    version: "1"
  }
}, (err,res) => {
  t.error(err)
  t.equal(res.statusCode, 400)
  t.match(res.payload, /uint8/)
})

fastify.inject({
  method: 'POST',
  url:'/',
  payload:{
    foo:'asd',
    version: "2"
  }
}, (err,res) => {
  t.error(err)
  t.equal(res.statusCode, 200)
  t.equal(res.json(), { ok: 1 })
})
```